pcap-ioplugin.c

```diff
@@ -0,0 +1,261 @@
+#endif
+
+#ifdef HAVE_OS_PROTO_H
+#include "os-proto.h"
+#endif
+
+/*
+ * Setting O_BINARY on DOS/Windows is a bit tricky
+ */
+#if defined(_WIN32)
+  #define SET_BINMODE(f)  _setmode(_fileno(f), _O_BINARY)
+#elif defined(MSDOS)
+  #if defined(__HIGHC__)
+  #define SET_BINMODE(f)  setmode(f, O_BINARY)
+  #else
+  #define SET_BINMODE(f)  setmode(fileno(f), O_BINARY)
+  #endif
+#else
+  #define SET_BINMODE(f)  (void)f
+#endif
+
+/*
+ * fopen's safe version on Windows.
+ */
+#ifdef _MSC_VER
+FILE *fopen_safe(const char *filename, const char* mode)
+{
+	FILE *fp = NULL;
+	errno_t errno;
+	errno = fopen_s(&fp, filename, mode);
+	if (errno == 0)
+		return fp;
+	else
+		return NULL;
+}
+#endif
+
+static FILE*
+stdio_open_read(const char *fname, char *errbuf)
+{
+	FILE *fp = NULL;
+
+	if (strcmp(fname, "-") == 0) {
+		fp = stdin;
+		SET_BINMODE(fp);
+	} else {
+		/*
+		 * "b" is supported as of C90, so *all* UN*Xes should
+		 * support it, even though it does nothing.  It's
+		 * required on Windows, as the file is a binary file
+		 * and must be written in binary mode.
+		 */
+		fp = fopen(fname, "rb");
+		if (fp == NULL) {
+			pcap_snprintf(errbuf, PCAP_ERRBUF_SIZE, "%s: fopen: %s", fname,
+				pcap_strerror(errno));
+			return (NULL);
+		}
+	}
+
+	return fp;
+}
+
+static FILE*
+stdio_open_write(const char *fname, char *errbuf)
+{
+	FILE *fp = NULL;
+
+	if (strcmp(fname, "-") == 0) {
+		fp = stdout;
+		SET_BINMODE(fp);
+	} else {
+		/*
+		 * "b" is supported as of C90, so *all* UN*Xes should
+		 * support it, even though it does nothing.  It's
+		 * required on Windows, as the file is a binary file
+		 * and must be written in binary mode.
+		 */
+		fp = fopen(fname, "wb");
+		if (fp == NULL) {
+			pcap_snprintf(errbuf, PCAP_ERRBUF_SIZE, "%s: fopen: %s", fname,
+				pcap_strerror(errno));
+			return (NULL);
+		}
+	}
+
+	return fp;
+}
+
+static const pcap_ioplugin_t* pcap_ioplugin_stdio() {
+	static pcap_ioplugin_t plugin = {
+		.open_read = stdio_open_read,
+		.open_write = stdio_open_write
+	};
+
+	return &plugin;
+}
+
+/*
+ * loads an I/O plugin with the given name
+ * (on UNIX, a .so shared library)
+ *
+ * NB: fails silently and falls back to existing uncompressed
+ *     stdio-based output if the plugin fails to load
+ */
+
+const pcap_ioplugin_t* pcap_ioplugin_init(const char *name)
+{
+	void *lib = NULL;
+	if (name == NULL) {
+		goto fail;
+	}
+
+#if HAVE_DLOPEN
+	lib = dlopen(name, RTLD_NOW);
+	if (lib != NULL) {
+		pcap_ioplugin_init_fn ioplugin_init = dlsym(lib, "ioplugin_init");
+		if (ioplugin_init == NULL) {
+			dlclose(lib);
+			goto fail;
+		} else {
+			return ioplugin_init();
+		}
+	}
+#endif /* HAVE_DLOPEN */
+
+fail:
+	return pcap_ioplugin_stdio();
+}
+
+struct file_entry {
+	FILE					*fp;
+	const void				*cookie;
+	struct file_entry		*next;
+};
+
+static struct {
+	struct file_entry		*head;
+} file_list = { NULL };
+
+static int registered_atexit = 0;
+static int running_atexit = 0;
+
+static void pcap_ioplugin_closeall()
```

Comment:
`static void pcap_ioplugin_closeall(void)` - C is still keeping around backwards compatibility with pre-prototype versions, so you have to put `void` to explicitly indicate that it takes no arguments.

Reply:
gah, too used to forgiving compilers! :P .  Resolved in 841b1b3, along with some other function declaration style changes to closer match the project style.